pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -882,19 +867,49 @@ func verifyStatus(ctx *domainContext, status *types.DomainStatus) {
 		// check if qemu processes has crashed
 		hasQemu := status.VirtualizationMode == types.HVM || status.VirtualizationMode == types.FML || status.IsContainer
 		if configActivate && status.Activated && hasQemu && !hyper.IsDeviceModelAlive(status.DomainId) {
-			errStr := fmt.Sprintf("verifyStatus(%s) qemu crashed",
+			//Have to check on the domain again before confirming qemu crash because of https://www.pivotaltracker.com/story/show/171962732
+			//Here we pass 3 sec wait before we verify if the domain is still alive.
+			//3 sec should be more than enough for the domain to kill qemu and shutdown itself.
+			log.Infof("verifyStatus(%s): qemu crashed or not running. Checking if the domain is still running",
 				status.Key())
-			log.Errorf(errStr)
-			status.LastErr = "qemu crashed - please restart application instance"
-			status.LastErrTime = time.Now()
-			status.Activated = false
-			status.State = types.HALTED
-			publishDomainStatus(ctx, status)
-			if err := hyper.Delete(status.DomainName, status.DomainId); err != nil {
-				log.Errorf("failed to delete domain: %s (%v)\n", status.DomainName, err)
+			_, err := getCurrentDomainID(ctx, configActivate, status, 3*time.Second)
+			if err == nil {
+				errStr := fmt.Sprintf("verifyStatus(%s) qemu crashed",
+					status.Key())
+				log.Errorf(errStr)
+				status.LastErr = "qemu crashed - please restart application instance"
+				status.LastErrTime = time.Now()
+				status.Activated = false
+				status.State = types.HALTED
+				publishDomainStatus(ctx, status)
+				if err := hyper.Delete(status.DomainName, status.DomainId); err != nil {
+					log.Errorf("failed to delete domain: %s (%v)\n", status.DomainName, err)
+				}
+			}
+		}
+	}
+}
+
+func getCurrentDomainID(ctx *domainContext, configActivate bool, domainStatus *types.DomainStatus, waitBeforeCheckSec time.Duration) (int, error) {
```

Comment:
This function has wierd side effects like calling publishDomainStatus.
If it is a function which the "get" name seems to imply, then it shouldn't have any side effects.

What's the point of splitting out this code? Makes it harder to understand.

Reply:
Changed the method name to verifyDomainCurrentStatus().
verifyStatus() now verifies domain twice. So rather than having the same code in 2 places, I separated it out as a method which I'm calling from both places.

This what I wanted to achieve:

```
Check on the domain:
  if the domain is dead:
    publish domain crashed
  else:
    Checking if the qemu is up
      if not:
        Check on the domain again: //To verify if the domain was in the middle of a shutdown
          wait for 3 sec 
          If the domain is up:
            publish qemu crash error
          else:
            publish domain crashed
```